post-processing/readme.qmd

```diff
@@ -0,0 +1,169 @@
+
+The plan is to have a small set of à la carte functions for specific operations. For example, let's look at a binary classification model with calibration and threshold optimization: 
+
+```{r}
+#| eval: false
+
+wflow_2 <- 
+  workflow() %>% 
+  add_formula(Class ~ .) %>% 
+  add_model(logistic_reg())
+
+wflow_individ <- 
+  wflow_2 %>% 
+  add_prob_calibration(cal_object) %>% 
+  add_prob_threshold(threshold = 0.7)
+```
+
+The new `add_*` functions would have `update_*()` and `remove_*()` analogs or overall functions that conduct those operations for all of the different post-processors. 
+
+Note that the order of operations matter. In the above example, the system should execute the calibration before threshold the probabilities since the calibration will change the data in a way that invalidates the new threshold. 
+
+However, the order that the operations are added to the worklflow should not matter. There are some rules that can be enacted (e.g. calibration before thresholding) but  the proposed solution is to have a prioritization scheme that can be, within reason, altered by the user. Each `add_*` function will have a `.priority` argument that has reasonable defaults (with one exception, see below).  
+
+### Current list of post-processors
+
+* Calibration of probability estimates to increase their probabilistic fidelity to the observed data
+
+```r
+# 'object' is a pre-made calibration object
+# Data Inputs: class probabilities
+# Data Outputs: class probabilities and recomputed class predictions
+add_prob_calibration(object, .priority = 1.0)
+```
+
+* For binary classification outcomes, there may be the need to optimize the probability threshold that produces hard class predictions.
+
+```r
+# Potentially tunable
+# Data Inputs: class probabilities
+# Data Outputs: class predictions
+add_prob_threshold(threshold = numeric(), .priority = 2.0)
+```
+
+* An addition of an equivocal zone where highly uncertain probability estimates where no hard class prediction is reported. 
+
+```r
+# Potentially tunable
+# Data Inputs: class probabilities
+# Data Outputs: class predictions
+add_cls_eq_zone(value = numeric(), threshold = numeric(), .priority = 3.0)
+```
+
+* Calibration of regression predictions to avoid regression to the mean or other systematic issues with models.
+
+```r
+# 'object' is a pre-made calibration object
+# Data Inputs: regression predictions
+# Data Outputs: regression predictions
+add_reg_calibration(object, .priority = 1.0)
+```
+
+* A general "mutate" operations where users can do on-the-fly operations using self-contained functions. For example,
+
+  - If a user log-transformed their data prior to model, then can exponentiate them here. 
+  - For regression models, the predictions can be thresholded to be within a specific range. 
+
+```r
+# User will have to always set the priority
+# Data Inputs: all predictions
+# Data Outputs: all predictions (only these columns are retained)
+add_post_mutate(..., .priority)
```

Comment:
I would consider giving all of these functions a common prefix that differentiates them from the other `add_*()` functions, like `add_post_*()`:

```r
add_post_calibration() # do you need prob vs reg calibration? can you just "figure it out"? can it be an argument?
add_post_threshold()
add_post_threshold_buffered()
add_post_mutate()
```

Reply:
Conflicted on this. I agree this would be nice for tab completion, but is inconsistent with the naming convention for preprocessors: `add_variables()`, `add_recipe()`, `add_formula()`.